### PR TITLE
fix Canal.startSyncer and position comparison 

### PR DESCRIPTION
**Pull request by bytewatch**
fix  two issue: 
* #339 , which may causes data loss.
* position comparison between different server id unexpectly, which may cause unexpect and stange situation.

### CODE REVIEW

**Comment by siddontang** [comment]
Thanks @bytewatch 

I will review it this weekend, too busy these days, sorry.

File: mysql/position.go [hunk 1]
```diff
@@ -29,5 +33,5 @@ func (p Position) Compare(o Position) int {
 }
 
 func (p Position) String() string {
-	return fmt.Sprintf("(%s, %d)", p.Name, p.Pos)
+	return fmt.Sprintf("%s:%d", p.Name, p.Pos)
```
Comment (siddontang) [inline mysql/position.go]: I prefer the original way :-)

Comment (bytewatch) [reply mysql/position.go]: I rollback it to old style now.

File: mysql/position.go [hunk 2]
```diff
@@ -6,11 +6,15 @@ import (
 
 // For binlog filename + position based replication
 type Position struct {
```
Comment (GregoryIan) [inline mysql/position.go]: I think `Position` is just binlog position,  if you need , I think it' better to create a new sturcture that contains server id and position

File: canal/sync.go [hunk 3]
```diff
@@ -31,9 +31,10 @@ func (c *Canal) startSyncer() (*replication.BinlogStreamer, error) {
 		log.Infof("start sync binlog at binlog file %v", pos)
 		return s, nil
 	} else {
-		s, err := c.syncer.StartSyncGTID(gset)
+		copyedGset, _ := mysql.ParseGTIDSet(c.cfg.Flavor, gset.String())
```
Comment (GregoryIan) [inline canal/sync.go]: maybe we should return a copy in `c.master.GTIDSet()` directly

**Comment by GregoryIan** [comment]
Rest LGTM

File: mysql/position.go (hunk 2 above)
Comment (bytewatch) [reply mysql/position.go]: But application that implements `canal.EventHandler` depends on `mysql.Position`, and some application need to know current MySQL master's server id,  to have a unique map between application's data:
```
 (server_id, name, pos) ---> data
``` 
Only this way , application can distinguish it's data when MySQL failover/switchover.

Comment (bytewatch) [reply mysql/position.go]: I append a commit 503c6b4 , add the logic about telling the current server_id to event handler (applications).

**Comment by siddontang** [comment]
PTAL @GregoryIan

File: canal/sync.go (hunk 3 above)
Comment (GregoryIan) [reply canal/sync.go]: you can  review  https://github.com/siddontang/go-mysql/blob/3fb4533b76294813dde21d94e80516828b0f977c/mysql/gtid.go#L17
there is a `Clone` function

File: mysql/position.go (hunk 2 above)
Comment (GregoryIan) [reply mysql/position.go]: what do you think about it @siddontang ?

Comment (bytewatch) [reply mysql/position.go]: I append a commit 12d3dda, fix comparison between different server_id in  `c.WaitUntilPos()` . I think a ha application need to care about MySQL failover/switchover.

Comment (siddontang) [reply mysql/position.go]: IMO, it is confused to add a ServerID in the Position structure, maybe we should use another way like checking server ID explicitly and separately.

Comment (bytewatch) [reply mysql/position.go]: Well, i drop the logic about server id. Now there are only logic about copy of GTIDSet. I call `Clone()` in `Canal.master.GTIDSet()` and `BinlogSyncer.getGtidSet()` to return copy of GTIDSet. PTAL.

**Comment by GregoryIan** [comment]
LGTM

**Pull request merged**